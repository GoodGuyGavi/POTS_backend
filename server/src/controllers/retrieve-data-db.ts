import { IDBModel } from '../commons/types';

const getAllDataDB = (db: IDBModel<any>) => async () => {
	return db.getAll();
};
const getByIDDB = (db: IDBModel<any>) => async id => {
	return db.getById(id);
};

// const getByDeliveryIdDB = (db: IDBModel<any>) => async id => {
//   return db.getByDeliverId(id);
// };

// const getAllByDriverDB = (db: IDBModel<any>) => async id => {
//   return db.getAllByDriver(id);
// };

export { getAllDataDB, getByIDDB };
