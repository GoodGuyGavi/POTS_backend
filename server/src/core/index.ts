import { createUCCreateSupplier } from './_usecases';

export { createUCCreateSupplier };
