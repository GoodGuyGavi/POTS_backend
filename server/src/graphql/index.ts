import { mergeTypes } from 'merge-graphql-schemas';

// import address from "./address";

const typeDefs = [
	//   address,
];

// NOTE: 2nd param is optional, and defaults to false
// Only use if you have defined the same type multiple times in
// different files and wish to attempt merging them together.
export default mergeTypes(typeDefs, { all: true });
