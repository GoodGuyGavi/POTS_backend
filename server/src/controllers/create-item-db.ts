import { IDBModel } from '../commons/types';
import { createUCCreateItem } from '../core/';

const createCreateItemDB = (db: IDBModel<any>) => async i => {
	const createItem = createUCCreateItem();
	const newItem = createItem(i);
	console.log('contorller', newItem);
	return db.insert(newItem);
};

export { createCreateItemDB };
