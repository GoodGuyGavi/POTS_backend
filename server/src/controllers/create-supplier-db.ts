import { IDBModel } from '../commons/types';
import { createUCCreateSupplier } from '../core/_usecases';

const createCreateSupplierDB = (db: IDBModel<any>) => async u => {
	const createSupplier = createUCCreateSupplier();
	const newSupplier = createSupplier(u);
	return db.insert(newSupplier);
};

export { createCreateSupplierDB };
