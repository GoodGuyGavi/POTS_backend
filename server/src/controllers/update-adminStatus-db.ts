import { IDBModel } from '../commons/types';
import { updateUCAdminStatus } from '../core/_usecases';

const updateAdminStatusByIDDB = (db: IDBModel<any>) => async data => {
  const a = await db.getById(data.id);
  const updateAdminStatus = updateUCAdminStatus();
  const newPurchaseOrder = updateAdminStatus(data, a);
  return db.updateById({ ...newPurchaseOrder, id: data.id });
};

export { updateAdminStatusByIDDB };
