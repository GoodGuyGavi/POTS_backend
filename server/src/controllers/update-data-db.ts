import { IDBModel } from '../commons/types';
import { updateUCUser } from '../core/_usecases';

const UpdateByIDDB = (db: IDBModel<any>) => async data => {
	return db.updateById(data);
};

export { UpdateByIDDB };
