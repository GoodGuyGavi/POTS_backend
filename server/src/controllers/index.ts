//User
import { createCreateUserDB } from './create-user-db';
import { DeleteRecordByIDDB } from './delete-data-db';
import { updateUserByIDDB } from './update-user-db';

//Supplier
import { createCreateSupplierDB } from './create-supplier-db';

import { getAllDataDB, getByIDDB } from './retrieve-data-db';

export { getAllDataDB, getByIDDB, createCreateUserDB, DeleteRecordByIDDB, updateUserByIDDB, createCreateSupplierDB };
