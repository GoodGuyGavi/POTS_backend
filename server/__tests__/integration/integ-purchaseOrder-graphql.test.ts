import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import mockData from '../mocks/mock-data';
import * as controllers from '../../src/controllers';
import { constructTestServer } from '../__utils';

const {
  getAllDataDB,
  getByIDDB,
  DeleteRecordByIDDB,
  getAllByItemDB,
  createCreatePurchaseOrderDB,
  createCreateItemDB,
  createCreateSupplierDB,
  createCreateScheduleLineDB,
  updatePurchaseOrderByIDDB,
} = controllers;

const purchaseOrderMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input, vendorAddress: 'A1', supplier: '1' };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.purchaseOrders.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.purchaseOrders;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.purchaseOrders.filter(filterData);
    return res[0] || null;
  },

  getAllBySupplierStatus: async id => {},
  getAllByItem: jest.fn(async () => {
    return mockData.purchaseOrders;
  }),
  getAllByScheduleLine: async id => {},
};

const itemMock = {
  insert: jest.fn(async input => {
    return {
      id: '1',
      ...input,
      deliveryAddress: 'A1',
    };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.items.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.users;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.items.filter(filterData);
    return res[0] || null;
  },
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: jest.fn(async id => {
    const filterData = data => {
      if (id.includes(data._id)) {
        return data;
      }
    };
    const res = mockData.scheduleLines
      .filter(filterData)
      .map(data => ({ ...data, id: data._id }));
    return res;
  }),
};

const supplierMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input, address: 'A1' };
  }),
  getAll: jest.fn(async () => {
    return mockData.suppliers;
  }),
  getById: jest.fn(async id => {
    const filterData = cust => {
      if (cust.id === id) {
        return cust;
      }
    };

    const res = mockData.suppliers.filter(filterData);
    return res[0] || null;
  }),

  deleteById: async id => {
    const filterData = cust => {
      if (cust.id === id) {
        return cust;
      }
    };

    const res = mockData.suppliers.filter(filterData);
    return res[0] || null;
  },

  updateById: jest.fn(async input => {
    return { ...input };
  }),
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
};

const scheduleLinesMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.scheduleLines;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);
    return res[0] || null;
  },
  getAllBySupplierStatus: jest.fn(async () => {
    return mockData.scheduleLines;
  }),
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
};

const addressMock = {
  insert: jest.fn(async input => {
    return { id: 'A1', ...input };
  }),
  getAll: jest.fn(async () => {
    return mockData.address;
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.address.filter(filterData);
    return res[0] || null;
  }),
  deleteById: async id => {},
  updateById: async id => {},
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
};

const { server }: any = constructTestServer({
  context: {
    createPurchaseOrder: createCreatePurchaseOrderDB(purchaseOrderMock),
    updatePurchaseOrderById: updatePurchaseOrderByIDDB(purchaseOrderMock),
    deletePurchaseOrderbyId: DeleteRecordByIDDB(purchaseOrderMock),
    getPurchaseOrderById: getByIDDB(purchaseOrderMock),
    getAllPurchaseOrders: getAllDataDB(purchaseOrderMock),
    getAllItemsByPurchaseOrder: getAllByItemDB(purchaseOrderMock),
    getItemById: getByIDDB(itemMock),
    createScheduleLine: createCreateScheduleLineDB(scheduleLinesMock),
    getAddressById: getByIDDB(addressMock),
    createItem: createCreateItemDB(itemMock),
    createSupplier: createCreateSupplierDB(supplierMock),
    getSupplierById: getByIDDB(supplierMock),
  },
});

describe('Tests', () => {
  //Queries
  it('should fetch all purchase orders', async () => {
    const PURCHASEORDER_ALL = gql`
      query {
        allPurchaseOrders {
          id
          purchaseOrderNo
          shipmentNo
          adminStatus
          supplierStatusHeader
          vendorAddress {
            id
            building_name
            street
            city
            state
            zip_code
          }
          supplier {
            id
            supplierNo
            supplierName
            tin
            contactNumber
            contactPerson
            address {
              id
              building_name
              street
              city
              state
              zip_code
            }
          }
          documentDate
          items {
            id
            itemNo
            description
            quantity
            uom
            unitPrice
            currency
            deliveryAddress {
              id
            }
            totalAmount
            supplierStatusItem
            scheduleLine {
              id
              quantity
              uom
              unitPrice
              totalAmount
              deliveryDateAndTime
              deliveryStatus {
                id
                status
                dateCreated
                timeCreated
              }
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({ query: PURCHASEORDER_ALL });

    expect(res).toMatchSnapshot();
  });

  it('should fetch one purchase order', async () => {
    const SINGLE_PURCHASEORDER = gql`
      query po($id: ID!) {
        purchaseOrder(id: $id) {
          id
          purchaseOrderNo
          shipmentNo
          adminStatus
          supplierStatusHeader
          vendorAddress {
            id
            building_name
            street
            city
            state
            zip_code
          }
          supplier {
            id
            supplierNo
            supplierName
            tin
            contactNumber
            contactPerson
            address {
              id
              building_name
              street
              city
              state
              zip_code
            }
          }
          documentDate
          items {
            id
            itemNo
            description
            quantity
            uom
            unitPrice
            currency
            deliveryAddress {
              id
            }
            totalAmount
            supplierStatusItem
            scheduleLine {
              id
              quantity
              uom
              unitPrice
              totalAmount
              deliveryDateAndTime
              deliveryStatus {
                id
                status
                dateCreated
                timeCreated
              }
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_PURCHASEORDER,
      variables: { id: '1' },
    });

    expect(res).toMatchSnapshot();
  });

  it('should error when no purchase order', async () => {
    const SINGLE_PURCHASEORDER = gql`
      query po($id: ID!) {
        purchaseOrder(id: $id) {
          id
          purchaseOrderNo
          shipmentNo
          adminStatus
          supplierStatusHeader
          vendorAddress {
            id
            building_name
            street
            city
            state
            zip_code
          }
          supplier {
            id
            supplierNo
            supplierName
            tin
            contactNumber
            contactPerson
            address {
              id
              building_name
              street
              city
              state
              zip_code
            }
          }
          documentDate
          items {
            id
            itemNo
            description
            quantity
            uom
            unitPrice
            currency
            deliveryAddress {
              id
            }
            totalAmount
            supplierStatusItem
            scheduleLine {
              id
              quantity
              uom
              unitPrice
              totalAmount
              deliveryDateAndTime
              deliveryStatus {
                id
                status
                dateCreated
                timeCreated
              }
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_PURCHASEORDER,
      variables: { id: '' },
    });

    expect(res).toMatchSnapshot();
  });

  //Mutations

  it('create a purchase order', async () => {
    const CREATE_PURCHASEORDER = gql`
      mutation createPO($purchaseOrder: PurchaseOrderInput) {
        createPurchaseOrder(purchaseOrder: $purchaseOrder) {
          purchaseOrderNo
          shipmentNo
          vendorAddress {
            id
            building_name
            street
            city
            state
            zip_code
          }
          supplier {
            id
            supplierNo
            supplierName
            tin
            contactNumber
            contactPerson
            address {
              id
              building_name
              street
              city
              state
              zip_code
            }
          }
          documentDate
          items {
            id
            itemNo
            description
            quantity
            uom
            unitPrice
            currency
            deliveryAddress {
              id
              building_name
              street
              city
              state
              zip_code
            }
            totalAmount
            supplierStatusItem
            scheduleLine {
              id
              quantity
              uom
              unitPrice
              totalAmount
              deliveryDateAndTime
              deliveryStatus {
                id
                status
                dateCreated
                timeCreated
              }
            }
          }
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CREATE_PURCHASEORDER,
      variables: {
        purchaseOrder: {
          purchaseOrderNo: '001',
          shipmentNo: '123',
          supplier: {
            supplierNo: '001',
            supplierName: 'Juan Dela Cruz',
            address: {
              building_name: '002',
              street: 'Elmer',
              city: 'Celadon',
              state: 'Johto',
              zip_code: '123',
            },
            contactPerson: 'Basil Valdez',
            contactNumber: '1234567',
            tin: '12345',
          },
          postingDate: '03/03/20200',
          documentDate: '03/03/20200',
          vendorAddress: {
            building_name: '002',
            street: 'Elmer',
            city: 'Celadon',
            state: 'Johto',
            zip_code: '123',
          },
          items: [
            {
              itemNo: '001',
              description: 'Beef',
              productId: '0001',
              quantity: 10,
              uom: 'kilograms',
              unitPrice: 1000,
              discount: 0.05,
              totalAmount: 10000,
              deliveryAddress: {
                building_name: 'building name',
                street: 'street',
                city: 'city',
                state: 'state',
                zip_code: 'zip_code',
              },
              scheduleLine: [
                {
                  quantity: 10,
                  uom: 'kilograms',
                  deliveryDateAndTime: 'February 25,2020 4:30PM',
                  unitPrice: 1000,
                  totalAmount: 10000,
                },
              ],
              currency: 'PHP',
            },
          ],
        },
      },
    });

    expect(res.errors).toBeUndefined();
    expect(purchaseOrderMock.insert.mock.calls.length).toBe(1);
    expect(res.data).toMatchObject({
      createPurchaseOrder: {
        _id: '1',
        purchaseOrderNo: '001',
        shipmentNo: '123',
        supplier: {
          id: '1',
          supplierNo: '001',
          supplierName: 'Juan Dela Cruz',
          address: {
            building_name: '002',
            street: 'Elmer',
            city: 'Celadon',
            state: 'Johto',
            zip_code: '123',
          },
          contactPerson: 'Basil Valdez',
          contactNumber: '1234567',
          tin: '12345',
        },
        postingDate: '03/03/20200',
        documentDate: '03/03/20200',
        vendorAddress: {
          building_name: '002',
          street: 'Elmer',
          city: 'Celadon',
          state: 'Johto',
          zip_code: '123',
        },
        items: [
          {
            id: '1',
            itemNo: '001',
            description: 'Beef',
            productId: '0001',
            quantity: 10,
            uom: 'kilograms',
            unitPrice: 1000,
            discount: 0.05,
            totalAmount: 10000,
            deliveryAddress: {
              building_name: 'building name',
              street: 'street',
              city: 'city',
              state: 'state',
              zip_code: 'zip_code',
            },
            scheduleLine: [
              {
                quantity: 10,
                uom: 'kilograms',
                deliveryDateAndTime: 'February 25,2020 4:30PM',
                unitPrice: 1000,
                totalAmount: 10000,
              },
            ],
            currency: 'PHP',
          },
        ],
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('update a purchase order', async () => {
    const UPDATE_PURCHASEORDER = gql`
      mutation i($item: UpdateItemInput!) {
        updateItem(item: $item) {
          id
          adminStatus
          supplierStatusHeader
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: UPDATE_PURCHASEORDER,
      variables: {
        item: {
          id: '1',
          adminStatus: 'Recieved',
          supplierStatusHeader: 'Delivered',
        },
      },
    });

    expect(res).toMatchSnapshot();
  });

  it('delete a puchase order', async () => {
    const DELETE_PURCHASEORDER = gql`
      mutation i($id: ID!) {
        deletePurchaseOrder(id: $id) {
          id
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: DELETE_PURCHASEORDER,
      variables: { id: '1' },
    });

    expect(res).toMatchSnapshot();
  });
});
