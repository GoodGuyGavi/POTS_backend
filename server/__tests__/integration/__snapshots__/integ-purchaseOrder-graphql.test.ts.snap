// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests delete a puchase order 1`] = `
Object {
  "data": Object {
    "deletePurchaseOrder": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should error when no purchase order 1`] = `
Object {
  "data": Object {
    "purchaseOrder": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch all purchase orders 1`] = `
Object {
  "data": Object {
    "allPurchaseOrders": Array [
      null,
      null,
    ],
  },
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field PurchaseOrder.id.],
    [GraphQLError: Cannot return null for non-nullable field PurchaseOrder.id.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch one purchase order 1`] = `
Object {
  "data": Object {
    "purchaseOrder": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests update a purchase order 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Cannot query field "adminStatus" on type "Item".],
    [ValidationError: Cannot query field "supplierStatusHeader" on type "Item". Did you mean "supplierStatusItem"?],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
